[Survey docs][3/n] - Add surveys code page (#6771)
PostHog/posthog.com: contents/docs/surveys/adding-surveys-code.mdx

---

> title: Adding survey code

Reviewer: This is slightly confusing because you need to "add code" for the popover surveys (just the snippet or library still counts as code.

Maybe "Implementing custom surveys"?

sidebar: Docs
showTitle: true
availability:
    free: none
    selfServe: full
    enterprise: full
---

> > **Note:** Adding code is **not required** for [popover surveys](/docs/surveys/creating-surveys#display-mode).

Reviewer: Same note as title, maybe implementing is a better word.

> If you're using PostHog's [survey API](/docs/surveys/creating-surveys#display-mode) to integrate a custom survey UI, you'll need to add your code before launching.

Reviewer: ```suggestion
If you're using PostHog's [survey API](/docs/surveys/creating-surveys#display-mode) to create a custom survey UI, you'll need to write fetch them and capture events before launching.
```

> There are 2 steps:

Reviewer: ```suggestion
```
Include the two steps in intro paragraph instead I think

> ## Step 1: Fetch your active surveys

Reviewer: ```suggestion
## Step 1: Fetch active surveys
```
Nit: They won't necessarily be "yours" but "users" so simpler to remove.

> To fetch the surveys which are currently enabled for your user, call `posthog.getActiveMatchingSurveys()`. 

Reviewer: ```suggestion
To fetch the surveys currently enabled for a user, call `posthog.getActiveMatchingSurveys()`. 
```

> Alternatively, if you wish to do survey targeting yourself, you can call `posthog.getSurveys()` to return a list of all surveys on PostHog.

Reviewer: ```suggestion
Alternatively, if you wish to do survey targeting yourself, you can call `posthog.getSurveys()` to return a list of all surveys from PostHog.
```

  "name": "Your survey name",
  "description": "Metadata describing your survey",
  "type": "api", // can be either "api" or "popover"
  "linked_flag_key": null, //Linked feature flag key, if any.
  "targeting_flag_key": "your_survey_targeting_flag_key",
  "questions": [
    {
      "type": "single_choice",
      "choices": [
        "Yes",
        "No"
      ],
      "question": "Are you enjoying PostHog?"
    }
  ],
  "conditions": null,
  "start_date": "2023-09-19T13:10:49.505000Z",
  "end_date": null
}]
```

Thus you can use show your user a survey using code like this:

```js-web
posthog.getActiveMatchingSurveys((surveys) => {
    // Make sure you filter for surveys that are of type `"api"`. 
    // This prevents accidentally showing popover surveys to your users.
    const filteredSurveys = surveys.filter(survey => survey.type === 'api'));
    if (filteredSurveys.length > 0) {
        // show the survey using your custom UI
        showSurvey(filteredSurveys[0]);
    }
})
```

> **Note:** If you're using a URL or selector targeting options, you'll need to poll `posthog.getActiveMatchingSurveys()` to get the most updated results.

## Step 2: Capture survey events

> You can capture user interactions with your surveys using the following 3 events:

Reviewer: A note to say "If you don't use these events, they won't show on the survey results page" would be nice.

```js-web

> // 1. When a user has seen a survey

Reviewer: ```suggestion
// 1. When a user is shown a survey
```

posthog.capture("survey shown", {
    $survey_id: {survey.id} // required
})

// 2. When a user has dismissed a survey
posthog.capture("survey dismissed", {
    $survey_id: {survey.id} // required
})

// 3. When a user has responded to a survey
posthog.capture("survey sent", {
    $survey_id: {survey.id} // required
    $survey_response: {survey_response} // required
})

> ```

Reviewer: Since we have these here, should we remove the `/docs/surveys/events` page? 

Someone using popovers doesn't really need to know them. Redirect to this section.

Author: Was planning to remove in next pr, but will just do it in this one now
